mbuild/compound.py

```diff
@@ -2236,6 +2349,200 @@ def _energy_minimize_openbabel(
                                 particle, particle.name
                             )
                         )
+        # initialize constraints
+        ob_constraints = openbabel.OBFFConstraints()
+
+        if distance_constraints is not None:
+            # check the dimensions of the array so that if a user passes
+            # a single constraint as a 1-D array,
+            # i.e., [(p1,p2), 2.0]  rather than [[(p1,p2), 2.0]],
+            # the for loop doesn't try to loop over the elements in the constraint
+            # as if they were each a constraint.
+            # if it is 1-d, just add it to a list.
+            temp1 = np.array(distance_constraints)
+            if len(temp1.shape) == 1:
+                distance_constraints = [distance_constraints]
+
+            for con_temp in distance_constraints:
+                p1 = con_temp[0][0]
+                p2 = con_temp[0][1]
+
+                if id(p1) == id(p2):
+                    raise MBuildError(
+                        f"Cannot create a constraint between a particle and itself: {p1} {p2} ."
+                    )
+                if len(p1.children) != 0 or len(p2.children) != 0:
+                    raise MBuildError(
+                        f"Constraints must correspond to individual atoms in Compounds."
+                    )
+
+                for part in [p1, p2]:
+                    if id(part) not in list(
+                        map(lambda x: id(x), self.successors())
+                    ):
+                        raise MBuildError(
+                            f"{part} is not a member of Compound {self}."
+                        )
+
+                pid_1 = -1
+                pid_2 = -1
+                index = 1  # openbabel indices start at 1
+                dist = con_temp[1] * 10.0  # convert to angstroms
+                # obenbabel uses angstroms, not nm,
+                # scale distance by 10.0
+                # indices_temp=[(0,0),dist*10.0]
+                for particle in self.particles():
+                    if id(p1) == id(particle):
+                        pid_1 = index
+                    if id(p2) == id(particle):
+                        pid_2 = index
+                    index += 1
```

Comment:
This can be a bottle neck if the Compound is huge. 
```suggestion
            # Before the for loop, create a dict of particle and its index
            particle_idx = {id(particle): idx for particle, idx in enumerate(self.particles())}
            
            # During the for loop
                dist = con_temp[1] * 10.0  # convert to angstroms
                # obenbabel uses angstroms, not nm,
                # scale distance by 10.0
                # indices_temp=[(0,0),dist*10.0]
                pid_1 = particle_idx[id(p1)] + 1 # openbabel indices start at 1 
                pid_2 = particle_idx[id(p2)] + 1
```

Reply:
Agreed, creating the maps right off the bat is best.